File: stations/system-monitor/scheduler.py
"""
System Monitor V2 Scheduler — launchd plist management for periodic reports.
"""

from __future__ import annotations

import json
import plistlib
import subprocess
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent
PLIST_DIR = Path.home() / "Library" / "LaunchAgents"
LOG_DIR = Path("~/.claude/data/system-monitor/logs").expanduser()

LABEL_PREFIX = "com.workshop.system-monitor"

PLIST_CONFIGS = {
    "weekly": {
        "label": f"{LABEL_PREFIX}-weekly",
        "calendar": {"Weekday": 1, "Hour": 5, "Minute": 0},  # Monday 05:00
    },
    "monthly": {
        "label": f"{LABEL_PREFIX}-monthly",
        "calendar": {"Day": 1, "Hour": 5, "Minute": 0},  # 1st of month 05:00
    },
}


class Scheduler:
    def __init__(self, config: dict | None = None):
        if config is None:
            config_path = SCRIPT_DIR / "config.json"
            config = json.loads(config_path.read_text()) if config_path.exists() else {}
        self.config = config
        schedule = config.get("schedule", {})

        # Override calendar intervals from config
        if "weekly" in schedule:
            w = schedule["weekly"]
            PLIST_CONFIGS["weekly"]["calendar"] = {
                "Weekday": w.get("day", 1),
                "Hour": w.get("hour", 5),
                "Minute": w.get("minute", 0),
            }
        if "monthly" in schedule:
            m = schedule["monthly"]
            PLIST_CONFIGS["monthly"]["calendar"] = {
                "Day": m.get("day", 1),
                "Hour": m.get("hour", 5),
                "Minute": m.get("minute", 0),
            }

    def _build_plist(self, report_type: str) -> dict:
        """Build launchd plist dict for a report type."""
        cfg = PLIST_CONFIGS[report_type]
        python = str(Path("~/.local/bin/python3").expanduser())
        main_script = str(SCRIPT_DIR / "__main__.py")

        LOG_DIR.mkdir(parents=True, exist_ok=True)

        return {
            "Label": cfg["label"],
            "ProgramArguments": [python, main_script, "report", f"--type={report_type}"],
            "StartCalendarInterval": cfg["calendar"],
            "WorkingDirectory": str(SCRIPT_DIR),
            "StandardOutPath": str(LOG_DIR / f"{report_type}-stdout.log"),
            "StandardErrorPath": str(LOG_DIR / f"{report_type}-stderr.log"),
            "EnvironmentVariables": {
                "PATH": "/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin",
                "HOME": str(Path.home()),
            },
        }

    def _plist_path(self, report_type: str) -> Path:
        return PLIST_DIR / f"{PLIST_CONFIGS[report_type]['label']}.plist"

    def install(self, report_type: str | None = None) -> list[str]:
        """Generate and load launchd plists. Returns list of installed labels."""
        PLIST_DIR.mkdir(parents=True, exist_ok=True)
        types = [report_type] if report_type else list(PLIST_CONFIGS.keys())
        installed = []

        for rt in types:
            plist_data = self._build_plist(rt)
            plist_path = self._plist_path(rt)
            label = PLIST_CONFIGS[rt]["label"]

            # Unload first if already loaded
            subprocess.run(
                ["launchctl", "unload", str(plist_path)],
                capture_output=True, timeout=10,
            )

            # Write plist
            with open(plist_path, "wb") as f:
                plistlib.dump(plist_data, f)

            # Load
            r = subprocess.run(
                ["launchctl", "load", str(plist_path)],
                capture_output=True, text=True, timeout=10,
            )
            if r.returncode == 0:
                installed.append(label)
            else:
                print(f"Warning: failed to load {label}: {r.stderr}", file=sys.stderr)

        return installed

    def uninstall(self, report_type: str | None = None) -> list[str]:
        """Unload and remove launchd plists. Returns list of uninstalled labels."""
        types = [report_type] if report_type else list(PLIST_CONFIGS.keys())
        uninstalled = []

        for rt in types:
            plist_path = self._plist_path(rt)
            label = PLIST_CONFIGS[rt]["label"]

            if plist_path.exists():
                subprocess.run(
                    ["launchctl", "unload", str(plist_path)],
                    capture_output=True, timeout=10,
                )
                plist_path.unlink()
                uninstalled.append(label)

        return uninstalled

    def status(self) -> dict:
        """Query schedule status for all report types."""
        result = {}
        list_out = subprocess.run(
            ["launchctl", "list"], capture_output=True, text=True, timeout=10,
        ).stdout

        for rt, cfg in PLIST_CONFIGS.items():
            label = cfg["label"]
            plist_path = self._plist_path(rt)
            loaded = label in list_out
            result[rt] = {
                "label": label,
                "plist_exists": plist_path.exists(),
                "loaded": loaded,
                "calendar": cfg["calendar"],
            }

        return result

    def run_now(self, report_type: str = "weekly") -> str:
        """Execute a report generation immediately. Returns report path."""
        from collector import collect_all, load_config
        from reporter import SystemReporter

        config = load_config()
        data = collect_all(config)
        reporter = SystemReporter(config)
        return reporter.generate(data, report_type)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="System Monitor Scheduler")
    parser.add_argument("action", choices=["install", "uninstall", "status", "run"],
                        help="Scheduler action")
    parser.add_argument("--type", choices=["weekly", "monthly"], default=None,
                        help="Report type (default: all)")
    args = parser.parse_args()

    scheduler = Scheduler()

    if args.action == "install":
        labels = scheduler.install(args.type)
        print(f"Installed: {', '.join(labels)}")
    elif args.action == "uninstall":
        labels = scheduler.uninstall(args.type)
        print(f"Uninstalled: {', '.join(labels)}")
    elif args.action == "status":
        print(json.dumps(scheduler.status(), indent=2))
    elif args.action == "run":
        path = scheduler.run_now(args.type or "weekly")
        print(f"Report: {path}")

File: stations/system-monitor/src/scheduler.rs
//! launchd plist management — enable/disable/restart/logs (already used by
//! the HTTP API) plus higher-level `ensure_weekly_plist` / `ensure_monthly_plist`
//! that materialise the periodic-report plists this station owns.
//!
//! Mirrors `stations/system-monitor/scheduler.py`: weekly = Mon 05:00,
//! monthly = day 1 05:00, both invoking `system-monitor reporter --type=…`.

use anyhow::{Context, Result};
use serde_json::{json, Value};
use std::path::{Path, PathBuf};

use crate::config::Settings;

const LABEL_PREFIX: &str = "com.workshop.system-monitor";
const LAUNCHCTL_TIMEOUT_SECS: u64 = 10;

// ─── Existing low-level helpers (kept as-is, used by api/mod.rs) ────────────────

pub async fn enable(label: &str) -> Result<Value> {
    let plist = format!("/Library/LaunchAgents/{label}.plist");
    let user_plist = std::env::var("HOME")
        .map(|h| format!("{h}/Library/LaunchAgents/{label}.plist"))
        .unwrap_or_default();
    let plist_path = if std::path::Path::new(&user_plist).exists() {
        user_plist
    } else {
        plist
    };
    let out = tokio::process::Command::new("launchctl")
        .args(["load", &plist_path])
        .output()
        .await?;
    Ok(json!({
        "ok": out.status.success(),
        "label": label,
        "stderr": String::from_utf8_lossy(&out.stderr),
    }))
}

pub async fn disable(label: &str) -> Result<Value> {
    let user_plist = std::env::var("HOME")
        .map(|h| format!("{h}/Library/LaunchAgents/{label}.plist"))
        .unwrap_or_default();
    let out = tokio::process::Command::new("launchctl")
        .args(["unload", &user_plist])
        .output()
        .await?;
    Ok(json!({
        "ok": out.status.success(),
        "label": label,
        "stderr": String::from_utf8_lossy(&out.stderr),
    }))
}

pub async fn restart(label: &str) -> Result<Value> {
    disable(label).await.ok();
    enable(label).await
}

pub async fn logs(label: &str, lines: usize) -> Result<Value> {
    let candidates = [
        format!("/tmp/{label}.log"),
        format!("/tmp/{label}.err"),
        format!("/var/log/{label}.log"),
    ];
    for c in &candidates {
        if std::path::Path::new(c).exists() {
            let out = tokio::process::Command::new("tail")
                .args(["-n", &lines.to_string(), c])
                .output()
                .await?;
            return Ok(json!({
                "label": label,
                "path": c,
                "lines": String::from_utf8_lossy(&out.stdout),
            }));
        }
    }
    Ok(json!({"label": label, "path": null, "lines": ""}))
}

// ─── High-level managed-plist API ───────────────────────────────────────────────

#[derive(Debug, Clone, Copy)]
pub enum ReportKind {
    Weekly,
    Monthly,
}

impl ReportKind {
    fn label(self) -> &'static str {
        match self {
            ReportKind::Weekly => "com.workshop.system-monitor-weekly",
            ReportKind::Monthly => "com.workshop.system-monitor-monthly",
        }
    }
    fn cli_arg(self) -> &'static str {
        match self {
            ReportKind::Weekly => "weekly",
            ReportKind::Monthly => "monthly",
        }
    }
    /// (key, value) pairs for `<StartCalendarInterval>`.
    fn calendar(self) -> Vec<(&'static str, i32)> {
        match self {
            // Monday 05:00
            ReportKind::Weekly => vec![("Weekday", 1), ("Hour", 5), ("Minute", 0)],
            // 1st of month 05:00
            ReportKind::Monthly => vec![("Day", 1), ("Hour", 5), ("Minute", 0)],
        }
    }
}

fn launch_agents_dir() -> PathBuf {
    PathBuf::from(std::env::var("HOME").unwrap_or_else(|_| "/tmp".into()))
        .join("Library/LaunchAgents")
}

fn plist_path_for(label: &str) -> PathBuf {
    launch_agents_dir().join(format!("{label}.plist"))
}

/// Enumerate plist files under `~/Library/LaunchAgents/` whose label matches
/// `com.workshop.system-monitor-*`. Returns one JSON entry per file.
pub fn list_managed_plists(_cfg: &Settings) -> Result<Vec<Value>> {
    let dir = launch_agents_dir();
    let mut out = Vec::new();
    let entries = match std::fs::read_dir(&dir) {
        Ok(e) => e,
        Err(_) => return Ok(out),
    };
    for entry in entries.flatten() {
        let path = entry.path();
        let Some(name) = path.file_name().and_then(|n| n.to_str()) else {
            continue;
        };
        if !name.starts_with(&format!("{LABEL_PREFIX}-")) || !name.ends_with(".plist") {
            continue;
        }
        let label = name.trim_end_matches(".plist").to_string();
        out.push(json!({
            "label": label,
            "path": path.display().to_string(),
            "exists": true,
        }));
    }
    Ok(out)
}

/// Write & `launchctl load` the weekly report plist.
pub fn ensure_weekly_plist(cfg: &Settings) -> Result<()> {
    ensure_plist(cfg, ReportKind::Weekly)
}

/// Write & `launchctl load` the monthly report plist.
pub fn ensure_monthly_plist(cfg: &Settings) -> Result<()> {
    ensure_plist(cfg, ReportKind::Monthly)
}

fn ensure_plist(cfg: &Settings, kind: ReportKind) -> Result<()> {
    let label = kind.label();
    let plist_path = plist_path_for(label);
    std::fs::create_dir_all(launch_agents_dir())
        .with_context(|| format!("create {}", launch_agents_dir().display()))?;

    let logs_dir = cfg.data_dir.join("logs");
    std::fs::create_dir_all(&logs_dir)
        .with_context(|| format!("create logs dir {}", logs_dir.display()))?;

    let exe = current_exe_path();
    let body = render_plist_xml(label, &exe, kind, &logs_dir);
    std::fs::write(&plist_path, body)
        .with_context(|| format!("write plist {}", plist_path.display()))?;

    // Best-effort unload-then-load so an existing plist is replaced.
    let _ = std::process::Command::new("launchctl")
        .args(["unload", plist_path.to_string_lossy().as_ref()])
        .output();
    let out = std::process::Command::new("launchctl")
        .args(["load", plist_path.to_string_lossy().as_ref()])
        .output()
        .with_context(|| format!("launchctl load {}", plist_path.display()))?;
    if !out.status.success() {
        anyhow::bail!(
            "launchctl load failed: {}",
            String::from_utf8_lossy(&out.stderr)
        );
    }
    let _ = LAUNCHCTL_TIMEOUT_SECS; // reserved for future timeout wiring
    Ok(())
}

fn current_exe_path() -> String {
    std::env::current_exe()
        .ok()
        .and_then(|p| p.to_str().map(|s| s.to_string()))
        .unwrap_or_else(|| "system-monitor".to_string())
}

fn render_plist_xml(label: &str, exe: &str, kind: ReportKind, logs_dir: &Path) -> String {
    let calendar_xml: String = kind
        .calendar()
        .into_iter()
        .map(|(k, v)| format!("        <key>{k}</key>\n        <integer>{v}</integer>\n"))
        .collect();
    let kind_arg = kind.cli_arg();
    let stdout = logs_dir.join(format!("{kind_arg}-stdout.log"));
    let stderr = logs_dir.join(format!("{kind_arg}-stderr.log"));
    let working = logs_dir.parent().unwrap_or(logs_dir);

    format!(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>{label}</string>
    <key>ProgramArguments</key>
    <array>
        <string>{exe}</string>
        <string>reporter</string>
        <string>--type={kind_arg}</string>
    </array>
    <key>StartCalendarInterval</key>
    <dict>
{calendar_xml}    </dict>
    <key>WorkingDirectory</key>
    <string>{working}</string>
    <key>StandardOutPath</key>
    <string>{stdout}</string>
    <key>StandardErrorPath</key>
    <string>{stderr}</string>
    <key>EnvironmentVariables</key>
    <dict>
        <key>PATH</key>
        <string>/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin</string>
    </dict>
</dict>
</plist>
"#,
        label = label,
        exe = exe,
        kind_arg = kind_arg,
        calendar_xml = calendar_xml,
        working = working.display(),
        stdout = stdout.display(),
        stderr = stderr.display(),
    )
}
